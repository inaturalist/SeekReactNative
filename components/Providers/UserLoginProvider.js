// @flow

import React, { useState, useEffect, useCallback } from "react";
import Realm from "realm";
import type { Node } from "react";

import { UserContext } from "../UserContext";
import { fetchAccessToken, fetchUserProfile } from "../../utility/loginHelpers";
import { setupChallenges } from "../../utility/challengeHelpers";
import realmConfig from "../../models";

type Props = {
  children: any
}

const UserLoginProvider = ( { children }: Props ): Node => {
  const [login, setLogin] = useState( null );
  const [userProfile, setUserProfile] = useState( null );

  const checkINatAdminStatus = ( profile ) => {
    if ( profile.roles.includes( "admin" ) ) {
      return true;
    }
    return false;
  };

  const updateSavedLogin = useCallback( async ( profile, token ): Promise<any> => {
    try {
      const realm = await Realm.open( realmConfig );
      const savedLogin = realm.objects( "LoginRealm" );

      if ( profile === null ) {
        return savedLogin;
      }

      if ( savedLogin.length === 0 || !savedLogin[0] ) {
        realm.write( ( ) => {
          realm.create( "LoginRealm", {
            loginToken: token,
            username: profile.login,
            profilePhoto: profile.icon_url,
            isAdmin: checkINatAdminStatus( profile )
          } );
        } );
      } else if (
        savedLogin[0].username !== profile.login
        || savedLogin[0].profilePhoto !== profile.icon_url
        || savedLogin[0].loginToken !== token
        || savedLogin[0].isAdmin !== checkINatAdminStatus( profile )
      ) {
        realm.write( ( ) => {
          savedLogin[0].username = profile.login;
          savedLogin[0].profilePhoto = profile.icon_url;
          savedLogin[0].isAdmin = checkINatAdminStatus( profile );
          savedLogin[0].loginToken = token;
        } );
      }
      return savedLogin;
    } catch ( e ) {
      console.log( e, "can't fetch saved login" );
    }
  }, [] );

  const deleteSavedLogin = async ( ) => {
    try {
      const realm = await Realm.open( realmConfig );
      const savedLogin = realm.objects( "LoginRealm" );
      realm.write( ( ) => {
        realm.delete( savedLogin );
      } );
    } catch ( e ) {
      console.log( e, "can't delete saved login" );
    }
  };

  const getLoggedIn = useCallback( async ( ) => {
    const token = await fetchAccessToken( );

    if ( !token ) {
      setupChallenges( false );
      await deleteSavedLogin( );
      setLogin( null );
    } else {
      // profile is null when no internet
      const profile = await fetchUserProfile( token );
      const savedLogin = await updateSavedLogin( profile, token );

      const profileObj = {
        login: savedLogin[0].username,
        // icon_url is higher resolution than icon
        icon: savedLogin[0].profilePhoto,
        isAdmin: savedLogin[0].isAdmin
      };
      setUserProfile( profileObj );
      setupChallenges( savedLogin[0].isAdmin );
      setLogin( savedLogin[0].loginToken );
    }
  }, [updateSavedLogin] );

  useEffect( ( ) => {
    getLoggedIn( );
  }, [getLoggedIn] );

  const updateLogin = ( ) => getLoggedIn( );

  const userContextValue = {
    login,
    updateLogin,
    userProfile
  };

  return (
    <UserContext.Provider value={userContextValue}>
      {children}
    </UserContext.Provider>
  );
};

export default UserLoginProvider;
