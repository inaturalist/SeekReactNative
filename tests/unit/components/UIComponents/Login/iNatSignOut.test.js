import React from "react";
import { render, screen, fireEvent } from "@testing-library/react-native";

import UserLoginProvider from "../../../../../components/Providers/UserLoginProvider";
import INatSignOut from "../../../../../components/UIComponents/Login/iNatSignOut";

function renderButton( user ) {
  return render(
    <UserLoginProvider value={user}>
      <INatSignOut />
    </UserLoginProvider>
  );
}

describe( "iNatSignOut", () => {
  test( "should render correctly", () => {
    renderButton();
    screen.findByText( "SIGN OUT OF INATURALIST" );
    expect( screen ).toMatchSnapshot();
  } );

  test( "press signs the user out", () => {
    const user = {
      login: "some_token",
      userProfile: {
        login: "some_name",
        icon: "some_photo"
      }
    };

    renderButton( user );
    const button = screen.getByText( "SIGN OUT OF INATURALIST" );
    fireEvent.press( button );
    // TODO: test that the user is signed out
  } );
} );
