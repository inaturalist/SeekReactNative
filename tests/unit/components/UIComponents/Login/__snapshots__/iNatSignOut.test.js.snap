// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iNatSignOut should render correctly 1`] = `
[
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        [
          {
            "color": "#000000",
            "fontFamily": "Whitney-Book",
            "fontSize": 16,
            "lineHeight": 21,
          },
          {
            "marginBottom": 33,
            "marginTop": 19,
          },
        ],
      ]
    }
  >
    Upload and keep track of your observations using the iNaturalist app or visiting inaturalist.org.

Thanks for being part of this community!
  </Text>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "#38976d",
        "borderRadius": 34,
        "height": 79,
        "justifyContent": "center",
        "maxWidth": 317,
        "opacity": 1,
        "width": 317,
      }
    }
    testID="greenButton"
  >
    <Text
      allowFontScaling={true}
      style={
        [
          {
            "textAlign": "left",
          },
          [
            {
              "color": "#ffffff",
              "fontFamily": "Whitney-Semibold",
              "fontSize": 18,
              "letterSpacing": 1,
              "paddingTop": 3,
              "textAlign": "center",
            },
            {
              "letterSpacing": 1,
            },
            {
              "fontSize": 18,
            },
          ],
        ]
      }
    >
      SIGN OUT OF INATURALIST
    </Text>
  </View>,
]
`;
